### PR TITLE
Introduce rejections

### PR DESCRIPTION
**Pull request by matszczygiel**
Introducing V5 wire protocol with rejection support. On older protocols, there is a fallback to cancelation.
Currently what I did for older protocols is that cancel is issued and if peer requests download after that it will receive the error message.
It can be changed so that we can emit an error upon calling `reject()` if our peer does not support rejection, what do you think?

The old <-> new protocol is actually not tested because TBH there is not an easy way to issue rejection from udrop. It would require proper manual testing on actual integration or to somehow devise a test framework for testing wire compatibility which we don't have at this moment.

There is one thing still missing. The temporaries are not deleted after rejection and in general, after the transfer is failed or terminated before all files are downloaded. It's a wider problem and deserves a PR on its own IMHO

### CODE REVIEW

File: drop-storage/src/types.rs
Comment (LukasPukenis) [inline drop-storage/src/types.rs:187]: The naming is inconsistent. The same goes for `Progress`. I think adding `File` prefix would be nice here

File: drop-transfer/src/ws/client/v2.rs
Comment (LukasPukenis) [inline drop-transfer/src/ws/client/v2.rs:126]: It's confusing to see which version works where. This is `v2` file so I would assume this is v2 logic in here however it's clearly v5 logic inside of this file.

Another point - rejection fails if the cancel fails. I think rejection should follow through fully locally despite any errors on the other end. This is only for this file as client/v4 and client/v5 have it without cancel.

it looks like this should be `on_reject` instead of `issue_reject`. TLDR: architecture is very confusing to an external reader, could you clarify things?

File: drop-transfer/src/ws/client/v4.rs
Comment (LukasPukenis) [inline drop-transfer/src/ws/client/v4.rs:206]: Could you explain this change?

**Review by LukasPukenis — COMMENTED** [review]
Rejection logic is for v5 only, I am not sure why v2 and v4 were updated at all. Overall rejection logic makes sense but there's a confusion in architecture. I've left a comment asking you to explain it

File: drop-transfer/src/ws/client/v4.rs
Comment (matszczygiel) [reply drop-transfer/src/ws/client/v4.rs:206]: I needed to use `.await` inside of lambda and that's the way to go

File: drop-transfer/src/ws/client/v2.rs
Comment (matszczygiel) [reply drop-transfer/src/ws/client/v2.rs:126]: Actually, V4 does not have rejections and it's the same logic.
I wrote in the opening comment that there are two approaches:
* fallback to cancel
* Return to the caller with the error

I picked the first one. Can switch to the second option quite painlessly

`on_x()` method generally contains the handler code for something that happened, e.g. when there is a download request you have the `on_download()` handler
`issue_x()` is when the caller requests something to happen, so, in this case, it is correctly named as `issue_reject()` since you cannot receive from peer rejection because the wire protocol does not support that

File: drop-storage/src/types.rs
Comment (matszczygiel) [reply drop-storage/src/types.rs]: Done

File: drop-transfer/src/ws/client/v2.rs
Comment (LukasPukenis) [reply drop-transfer/src/ws/client/v2.rs:126]: I think I got it. `client/vN.rs` are client code when communicating with that version, for example, `client/v2.rs` might still be used if we communicate with v2 server? In that case it would make sense

Comment (matszczygiel) [reply drop-transfer/src/ws/client/v2.rs:126]: Yes, `client/v2.rs` is a client implementation for communicating with the V2 protocol. The same goes for the server.

So I presented in this PR the fallback for cancellation in the case of V2 and V4 protocols.
Do you think that's OK? We can switch to reporting errors to the caller, effectively disabling this functionality for communicating with legacy peers

Comment (matszczygiel) [reply drop-transfer/src/ws/client/v2.rs:126]: The conclusion is that for this moment we fallback to cancel

**Review by LukasPukenis — DISMISSED** [review]
I've looked through test cases and API, both make sense. The tests are very nice. I however hope for the best on the protocol level as there are a lot of changes but at a glance, they look good

@Lipt0nas should take a look as well

**Comment by matszczygiel** [comment]
The PR was approved before minor changes, so I'm merging this one

**Pull request merged**
